composer/profiler/performance_analysis_trace_handler.py

```diff
@@ -0,0 +1,114 @@
+        # we also need to put a floor millisecond threshold to account for python overhead in
+        # yielding a batch, which could potentially exceed 1% for very small models.
+        self.DATALOADER_PERCENTAGE_THRESHOLD = 0.01
+        self.DATALOADER_TIME_THRESHOLD = 1.0e6  # 1 millisecond or 1e6 nanoseconds
+
+    def reset_wait_times(self, state: State, logger: Logger) -> None:
+        """Reset wait times if algorithms have changed in response to bottleneck.
+
+        This function should be called to reset the trace handler if any changes have been made to the
+        training loop which may change dataloader bottlenecked status.
+        """
+        del state, logger
+        self.dataloader_wait_times = {}
+        self.batch_wait_times = {}
+
+    def epoch_start(self, state: State, logger: Logger) -> None:
+        self.reset_wait_times(state, logger)
+
+    def _record_wait_time(
+        self,
+        is_start: bool,
+        wait_times: Dict[int, Any],
+        wait_key: int,
+        wall_clock_time_ns: int,
+    ) -> None:
+        if is_start:
+            self.last_start_time = wall_clock_time_ns
+        elif self.last_start_time:
+            latency = wall_clock_time_ns - self.last_start_time
+            wait_times[wait_key] = latency
+
+    def process_duration_event(
+        self,
+        name: str,
+        categories: Union[List[str], Tuple[str, ...]],
+        is_start: bool,
+        timestamp: Timestamp,
+        wall_clock_time_ns: int,
+        action: ProfilerAction,
+    ) -> None:
+        del name
+        if 'dataloader' in categories:
+            self._record_wait_time(is_start, self.dataloader_wait_times, timestamp.batch.value, wall_clock_time_ns)
+        # Check for both [before/after]_train_batch and PerformanceAnalyzerTraceHandler so it only
+        # fires once on [before/after]_train_batch
+        elif 'before_train_batch' in categories and 'PerformanceAnalyzerTraceHandler' in categories:
+            self._record_wait_time(True, self.batch_wait_times, timestamp.batch.value, wall_clock_time_ns)
+        elif 'after_train_batch' in categories and 'PerformanceAnalyzerTraceHandler' in categories:
+            self._record_wait_time(False, self.batch_wait_times, timestamp.batch.value, wall_clock_time_ns)
+            # Only check for bottleneck on end of batch
+            if not is_start and action == ProfilerAction.ACTIVE_AND_SAVE and self.is_dataloader_bottlenecked():
+                dataloader_time = self._average_dataloader_wait_time()
+                batch_time = self._average_batch_wait_time()
+                # Log percentage overhead and additional time in milliseconds
+                percentage = format(dataloader_time / batch_time * 100, '.2f')
+                additional_latency = format(dataloader_time / 1e6, '.3f')
+                warnings.warn(f'The current training run is dataloader bottlenecked, adding {additional_latency}ms and '
```

Comment:
nit: the formating can go directly in the format string below; no need to call `format` manually -- e.g.

 ```suggestion
                 warnings.warn(f'The current training run is dataloader bottlenecked, adding {additional_latency:.3f}ms and '
```